File: src/notices/orchestrator.ts
import { Collection } from 'mongodb';
import type { Logger } from 'pino';
import { getDb } from '../shared/db.js';
import { Fetcher } from '../shared/fetcher.js';
import { Notice, NoticeListItem, buildNotice } from './normalizer.js';
import { cleanHtml } from './cleanHtml.js';
import {
  ensureIndexes, findExistingMeta, hasChanged, shouldContinue,
  upsertNotice, bulkTouchNotices, findNullContent,
  type ExistingMeta,
} from './dedup.js';
import { SkkuStandardStrategy } from './strategies/skku-standard.js';
import { WordPressApiStrategy } from './strategies/wordpress-api.js';
import { SkkumedAspStrategy } from './strategies/skkumed-asp.js';
import { JspDormStrategy } from './strategies/jsp-dorm.js';
import { CustomPhpStrategy } from './strategies/custom-php.js';
import { GnuboardStrategy } from './strategies/gnuboard.js';
import { GnuboardCustomStrategy } from './strategies/gnuboard-custom.js';
import type { DepartmentConfig, CrawlStrategy } from './types.js';
import { randomUUID } from 'node:crypto';
import baseLogger from '../shared/logger.js';

export interface CrawlOptions {
  incremental: boolean;
  maxPages?: number;
  delayMs?: number;
  deptFilter?: string;
}

interface DeptResult {
  deptId: string;
  deptName: string;
  inserted: number;
  updated: number;
  skipped: number;
  errors: number;
  durationMs: number;
}

const strategyMap: Record<string, (fetcher: Fetcher) => CrawlStrategy> = {
  'skku-standard': (fetcher) => new SkkuStandardStrategy(fetcher),
  'wordpress-api': (fetcher) => new WordPressApiStrategy(fetcher),
  'skkumed-asp': (fetcher) => new SkkumedAspStrategy(fetcher),
  'jsp-dorm': (fetcher) => new JspDormStrategy(fetcher),
  'custom-php': (fetcher) => new CustomPhpStrategy(fetcher),
  'gnuboard': (fetcher) => new GnuboardStrategy(fetcher),
  'gnuboard-custom': (fetcher) => new GnuboardCustomStrategy(fetcher),
};

export async function runCrawl(
  departments: DepartmentConfig[],
  options: CrawlOptions
): Promise<DeptResult[]> {
  const crawlId = randomUUID().slice(0, 8);
  const logger = baseLogger.child({ crawlId });

  const db = await getDb();
  const collection = db.collection<Notice>('notices');
  await ensureIndexes(collection);

  const pLimit = (await import('p-limit')).default;
  const limit = pLimit(5);

  const fetcher = new Fetcher({ delayMs: options.delayMs });

  const filtered = options.deptFilter
    ? departments.filter((d) => d.id === options.deptFilter)
    : departments;

  if (filtered.length === 0) {
    logger.warn({ deptFilter: options.deptFilter }, 'No matching departments found');
    return [];
  }

  const tasks = filtered.map((dept) =>
    limit(() => crawlDepartment(dept, collection, fetcher, options, logger))
  );

  const results = await Promise.allSettled(tasks);

  const deptResults: DeptResult[] = [];
  for (const r of results) {
    if (r.status === 'fulfilled') {
      deptResults.push(r.value);
    } else {
      logger.error({ error: r.reason }, 'Department crawl failed');
    }
  }

  const totalInserted = deptResults.reduce((s, r) => s + r.inserted, 0);
  const totalUpdated = deptResults.reduce((s, r) => s + r.updated, 0);
  const totalSkipped = deptResults.reduce((s, r) => s + r.skipped, 0);
  const totalErrors = deptResults.reduce((s, r) => s + r.errors, 0);
  logger.info(
    { departments: deptResults.length, totalInserted, totalUpdated, totalSkipped, totalErrors },
    'Crawl completed'
  );

  return deptResults;
}

async function crawlDepartment(
  dept: DepartmentConfig,
  collection: Collection<Notice>,
  fetcher: Fetcher,
  options: CrawlOptions,
  logger: Logger
): Promise<DeptResult> {
  const start = Date.now();
  const strategyFactory = strategyMap[dept.strategy];
  if (!strategyFactory) {
    throw new Error(`Unknown strategy: ${dept.strategy}`);
  }

  const strategy = strategyFactory(fetcher);
  const result: DeptResult = {
    deptId: dept.id,
    deptName: dept.name,
    inserted: 0,
    updated: 0,
    skipped: 0,
    errors: 0,
    durationMs: 0,
  };

  logger.info({ deptId: dept.id, deptName: dept.name }, 'Starting department crawl');

  // Re-crawl previously failed detail pages (content: null)
  const nullContentRefs = await findNullContent(collection, dept.id);
  if (nullContentRefs.length > 0) {
    logger.info({ count: nullContentRefs.length, deptId: dept.id }, 'Re-crawling null content articles');
    for (const ref of nullContentRefs) {
      const detail = await strategy.crawlDetail(ref, dept);
      if (detail) {
        await collection.updateOne(
          { articleNo: ref.articleNo, sourceDeptId: dept.id },
          {
            $set: {
              content: detail.content,
              contentText: detail.contentText,
              cleanHtml: cleanHtml(detail.content, dept.baseUrl),
              attachments: detail.attachments,
              crawledAt: new Date(),
            },
          }
        );
        result.updated++;
      }
    }
  }

  // Crawl list pages
  const maxPages = options.maxPages ?? (options.incremental ? 100 : 2500);
  let page = 0;

  while (page < maxPages) {
    let listItems: NoticeListItem[];
    try {
      listItems = await strategy.crawlList(dept, page);
    } catch (err) {
      logger.error({ deptId: dept.id, page, error: (err as Error).message }, 'Failed to fetch list page');
      result.errors++;
      break;
    }

    if (listItems.length === 0) {
      logger.info({ deptId: dept.id, page }, 'Empty list page — end of notices');
      break;
    }

    const isFirstPage = page === 0;

    if (options.incremental) {
      const articleNos = listItems.map((item) => item.articleNo);
      const existingMeta = await findExistingMeta(collection, dept.id, articleNos);
      const allKnown = !shouldContinue(listItems, existingMeta);

      if (!isFirstPage && allKnown) {
        logger.info({ page }, 'All items already in DB — stopping pagination');
        break;
      }

      if (isFirstPage && allKnown) {
        logger.info('All items on page 1 already in DB — early stop');
        await processPageSmart(listItems, existingMeta, strategy, dept, collection, result, logger);
        break;
      }

      await processPageSmart(listItems, existingMeta, strategy, dept, collection, result, logger);
    } else {
      await processPageFull(listItems, strategy, dept, collection, result, logger);
    }

    page++;
  }

  result.durationMs = Date.now() - start;
  logger.info(
    { ...result },
    'Department crawl finished'
  );
  return result;
}

async function processPageSmart(
  listItems: NoticeListItem[],
  existingMeta: Map<number, ExistingMeta>,
  strategy: CrawlStrategy,
  dept: DepartmentConfig,
  collection: Collection<Notice>,
  result: DeptResult,
  logger: Logger
): Promise<void> {
  const toTouch: { articleNo: number; sourceDeptId: string; views: number }[] = [];

  for (const item of listItems) {
    try {
      const existing = existingMeta.get(item.articleNo);

      if (existing && !hasChanged(item, existing)) {
        toTouch.push({ articleNo: item.articleNo, sourceDeptId: dept.id, views: item.views });
        result.skipped++;
        continue;
      }

      if (existing) {
        logger.info(
          { articleNo: item.articleNo, oldTitle: existing.title, newTitle: item.title },
          'Change detected — re-fetching detail'
        );
      }

      const detail = await strategy.crawlDetail({ articleNo: item.articleNo, detailPath: item.detailPath }, dept);
      const notice = buildNotice(item, detail, {
        department: dept.name,
        sourceDeptId: dept.id,
        baseUrl: dept.baseUrl,
      });

      const action = await upsertNotice(collection, notice);
      if (action === 'inserted') result.inserted++;
      else result.updated++;
    } catch (err) {
      logger.error(
        { articleNo: item.articleNo, error: (err as Error).message },
        'Failed to process article'
      );
      result.errors++;
    }
  }

  if (toTouch.length > 0) {
    await bulkTouchNotices(collection, toTouch);
  }
}

async function processPageFull(
  listItems: NoticeListItem[],
  strategy: CrawlStrategy,
  dept: DepartmentConfig,
  collection: Collection<Notice>,
  result: DeptResult,
  logger: Logger
): Promise<void> {
  for (const item of listItems) {
    try {
      const detail = await strategy.crawlDetail({ articleNo: item.articleNo, detailPath: item.detailPath }, dept);
      const notice = buildNotice(item, detail, {
        department: dept.name,
        sourceDeptId: dept.id,
        baseUrl: dept.baseUrl,
      });

      const action = await upsertNotice(collection, notice);
      if (action === 'inserted') result.inserted++;
      else result.updated++;
    } catch (err) {
      logger.error(
        { articleNo: item.articleNo, error: (err as Error).message },
        'Failed to process article'
      );
      result.errors++;
    }
  }
}

File: py/src/skkuverse_crawler/notices/orchestrator.py
from __future__ import annotations

import asyncio
import time
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

from ..shared.db import get_db
from ..shared.fetcher import Fetcher
from ..shared.html_cleaner import clean_html, normalize_content_urls
from ..shared.logger import get_logger
from .dedup import (
    bulk_touch_notices,
    ensure_indexes,
    find_existing_meta,
    find_null_content,
    has_changed,
    should_continue,
    update_with_history,
    upsert_notice,
)
from .constants import SERVICE_START_DATE
from .hashing import compute_content_hash
from .image_verifier import verify_notice_images
from .models import Notice, NoticeListItem
from .normalizer import build_notice
from .strategies.skku_standard import SkkuStandardStrategy
from .strategies.wordpress_api import WordPressApiStrategy
from .strategies.skkumed_asp import SkkumedAspStrategy
from .strategies.jsp_dorm import JspDormStrategy
from .strategies.custom_php import CustomPhpStrategy
from .strategies.gnuboard import GnuboardStrategy
from .strategies.gnuboard_custom import GnuboardCustomStrategy


_MAX_CONTENT_BYTES = 5 * 1024 * 1024  # 5MB


@dataclass
class CrawlOptions:
    incremental: bool = True
    max_pages: int | None = None
    delay_ms: int | None = None
    dept_filter: tuple[str, ...] | None = None


@dataclass
class DeptResult:
    dept_id: str = ""
    dept_name: str = ""
    inserted: int = 0
    updated: int = 0
    skipped: int = 0
    errors: int = 0
    duration_ms: int = 0


STRATEGY_MAP: dict[str, type] = {
    "skku-standard": SkkuStandardStrategy,
    "wordpress-api": WordPressApiStrategy,
    "skkumed-asp": SkkumedAspStrategy,
    "jsp-dorm": JspDormStrategy,
    "custom-php": CustomPhpStrategy,
    "gnuboard": GnuboardStrategy,
    "gnuboard-custom": GnuboardCustomStrategy,
}


async def run_crawl(
    departments: list[dict[str, Any]],
    options: CrawlOptions,
) -> list[DeptResult]:
    crawl_id = uuid.uuid4().hex[:8]
    logger = get_logger("orchestrator", crawl_id=crawl_id)

    db = await get_db()
    collection = db["notices"]
    await ensure_indexes(collection)

    fetcher = Fetcher(delay_ms=options.delay_ms or 500)

    if options.dept_filter:
        valid_ids = {d["id"] for d in departments}
        unknown = [did for did in options.dept_filter if did not in valid_ids]
        if unknown:
            raise ValueError(
                f"Unknown department ID(s) in CRAWL_DEPT_FILTER: {unknown}. "
                f"Check departments.json for valid IDs."
            )
        filtered = [d for d in departments if d["id"] in options.dept_filter]
    else:
        filtered = departments

    if not filtered:
        logger.warning("no_matching_departments", dept_filter=options.dept_filter)
        return []

    sem = asyncio.Semaphore(5)
    results: list[DeptResult] = []

    async def crawl_with_sem(dept: dict) -> DeptResult:
        async with sem:
            return await _crawl_department(dept, collection, fetcher, options, logger)

    tasks = [crawl_with_sem(dept) for dept in filtered]
    settled = await asyncio.gather(*tasks, return_exceptions=True)

    for r in settled:
        if isinstance(r, DeptResult):
            results.append(r)
        else:
            logger.error("department_crawl_failed", error=str(r))

    total_inserted = sum(r.inserted for r in results)
    total_updated = sum(r.updated for r in results)
    total_skipped = sum(r.skipped for r in results)
    total_errors = sum(r.errors for r in results)
    logger.info(
        "crawl_completed",
        departments=len(results),
        total_inserted=total_inserted,
        total_updated=total_updated,
        total_skipped=total_skipped,
        total_errors=total_errors,
    )

    await fetcher.close()
    return results


async def _crawl_department(
    dept: dict[str, Any],
    collection: Any,
    fetcher: Fetcher,
    options: CrawlOptions,
    logger: Any,
) -> DeptResult:
    start = time.monotonic()
    strategy_cls = STRATEGY_MAP.get(dept["strategy"])
    if not strategy_cls:
        raise ValueError(f"Unknown strategy: {dept['strategy']}")

    strategy = strategy_cls(fetcher)
    result = DeptResult(dept_id=dept["id"], dept_name=dept["name"])

    logger.info("starting_department_crawl", dept_id=dept["id"], dept_name=dept["name"])

    # Re-crawl null content
    null_refs = await find_null_content(collection, dept["id"])
    if null_refs:
        logger.info("recrawling_null_content", count=len(null_refs), dept_id=dept["id"])
        for ref in null_refs:
            detail = await strategy.crawl_detail(
                {"articleNo": ref["articleNo"], "detailPath": ref["detailPath"]}, dept
            )
            if detail:
                cleaned = clean_html(detail.content, dept["baseUrl"])
                raw_content = normalize_content_urls(detail.content, dept["baseUrl"])
                if cleaned and len(cleaned.encode()) > _MAX_CONTENT_BYTES:
                    logger.warning(
                        "oversized_content_dropped",
                        articleNo=ref["articleNo"],
                        dept=dept["id"],
                        size=len(cleaned.encode()),
                    )
                    cleaned = None
                    raw_content = None
                await collection.update_one(
                    {"articleNo": ref["articleNo"], "sourceDeptId": dept["id"]},
                    {"$set": {
                        "content": raw_content,
                        "contentText": detail.contentText,
                        "cleanHtml": cleaned,
                        "contentHash": compute_content_hash(cleaned),
                        "attachments": detail.attachments,
                        "crawledAt": datetime.now(timezone.utc),
                    }},
                )
                result.updated += 1

    # Crawl list pages
    max_pages = options.max_pages or (100 if options.incremental else 2500)
    page = 0

    while page < max_pages:
        try:
            list_items = await strategy.crawl_list(dept, page)
        except Exception as exc:
            logger.error("list_fetch_failed", dept_id=dept["id"], page=page, error=str(exc))
            result.errors += 1
            break

        if not list_items:
            logger.info("empty_list_page", dept_id=dept["id"], page=page)
            break

        if all(item.date and item.date < SERVICE_START_DATE for item in list_items):
            logger.info("floor_date_stopping", page=page, dept_id=dept["id"])
            break

        is_first_page = page == 0

        if options.incremental:
            article_nos = [item.articleNo for item in list_items]
            existing_meta = await find_existing_meta(collection, dept["id"], article_nos)
            all_known = not should_continue(list_items, existing_meta)

            if not is_first_page and all_known:
                logger.info("all_known_stopping", page=page)
                break

            if is_first_page and all_known:
                logger.info("all_known_first_page_early_stop")
                await _process_page_smart(
                    list_items, existing_meta, strategy, dept, collection, result, logger
                )
                break

            await _process_page_smart(
                list_items, existing_meta, strategy, dept, collection, result, logger
            )
        else:
            await _process_page_full(list_items, strategy, dept, collection, result, logger)

        page += 1

    result.duration_ms = int((time.monotonic() - start) * 1000)
    logger.info(
        "department_crawl_finished",
        dept_id=result.dept_id,
        inserted=result.inserted,
        updated=result.updated,
        skipped=result.skipped,
        errors=result.errors,
        duration_ms=result.duration_ms,
    )
    return result


async def _verify_and_log_images(notice: Notice, dept_id: str) -> None:
    """Best-effort image verification — never raises, only logs."""
    try:
        result = await verify_notice_images(notice.content, notice.sourceUrl)
        if result.broken:
            logger.warning(
                "broken_notice_images",
                articleNo=notice.articleNo,
                dept_id=dept_id,
                checked=result.checked,
                broken_count=len(result.broken),
                broken=result.broken[:5],  # cap log payload
            )
    except Exception as exc:
        logger.warning(
            "image_verify_failed",
            articleNo=notice.articleNo,
            dept_id=dept_id,
            error=str(exc),
        )


async def _process_page_smart(
    list_items: list[NoticeListItem],
    existing_meta: dict[int, dict[str, Any]],
    strategy: Any,
    dept: dict[str, Any],
    collection: Any,
    result: DeptResult,
    logger: Any,
) -> None:
    to_touch: list[dict[str, Any]] = []

    for item in list_items:
        try:
            if item.date and item.date < SERVICE_START_DATE:
                result.skipped += 1
                continue

            existing = existing_meta.get(item.articleNo)

            if existing and not has_changed(item, existing):
                to_touch.append({
                    "articleNo": item.articleNo,
                    "sourceDeptId": dept["id"],
                    "views": item.views,
                })
                result.skipped += 1
                continue

            detail = await strategy.crawl_detail(
                {"articleNo": item.articleNo, "detailPath": item.detailPath}, dept
            )
            notice = build_notice(
                item, detail,
                department=dept["name"],
                source_dept_id=dept["id"],
                base_url=dept["baseUrl"],
            )
            await _verify_and_log_images(notice, dept["id"])

            if not existing:
                action = await upsert_notice(collection, notice)
                if action == "inserted":
                    result.inserted += 1
                else:
                    result.updated += 1
            else:
                logger.info(
                    "change_detected",
                    articleNo=item.articleNo,
                    old_title=existing["title"],
                    new_title=item.title,
                )
                old_hash = existing.get("contentHash")
                new_hash = notice.contentHash
                edit_entry = {
                    "detectedAt": datetime.now(timezone.utc),
                    "oldHash": old_hash,
                    "newHash": new_hash,
                    "oldTitle": existing["title"],
                    "newTitle": item.title,
                    "titleChanged": existing["title"] != item.title,
                    "contentChanged": old_hash is not None and old_hash != new_hash,
                    "source": "tier1",
                }
                await update_with_history(collection, notice, edit_entry)
                result.updated += 1

        except Exception as exc:
            logger.error("process_article_failed", articleNo=item.articleNo, error=str(exc))
            result.errors += 1

    if to_touch:
        await bulk_touch_notices(collection, to_touch)


async def _process_page_full(
    list_items: list[NoticeListItem],
    strategy: Any,
    dept: dict[str, Any],
    collection: Any,
    result: DeptResult,
    logger: Any,
) -> None:
    for item in list_items:
        try:
            if item.date and item.date < SERVICE_START_DATE:
                result.skipped += 1
                continue

            detail = await strategy.crawl_detail(
                {"articleNo": item.articleNo, "detailPath": item.detailPath}, dept
            )
            notice = build_notice(
                item, detail,
                department=dept["name"],
                source_dept_id=dept["id"],
                base_url=dept["baseUrl"],
            )
            await _verify_and_log_images(notice, dept["id"])
            action = await upsert_notice(collection, notice)
            if action == "inserted":
                result.inserted += 1
            else:
                result.updated += 1

        except Exception as exc:
            logger.error("process_article_failed", articleNo=item.articleNo, error=str(exc))
            result.errors += 1
